polar-core/src/partial/simplify.rs

```diff
@@ -1,270 +1,361 @@
+                {
+                    right.clone()
+                }
+                (_, Value::Variable(r)) | (_, Value::RestVariable(r))
+                    if r == &this && left.is_ground() =>
+                {
+                    left.clone()
+                }
+                _ => term,
+            }
+        }
+        _ => term,
+    }
+}
+
+pub fn simplify_partial(var: &Symbol, term: Term) -> Term {
+    let mut simplifier = Simplifier::new(var.clone());
+    let simplified = simplifier.simplify_partial(term);
+    let simplified = simplify_trivial_constraint(var.clone(), simplified);
+    if matches!(simplified.value(), Value::Expression(e) if e.operator != Operator::And) {
+        op!(And, simplified).into_term()
+    } else {
+        simplified
+    }
+}
 
 /// Simplify the values of the bindings to be returned to the host language.
 ///
 /// - For partials, simplify the constraint expressions.
-/// - For non-partials, deep deref.
-pub fn simplify_bindings(bindings: Bindings) -> Bindings {
-    bindings
-        .into_iter()
-        .map(|(var, value)| match value.value() {
-            Value::Partial(_) => {
-                let simplified = simplify_partial(value);
-                assert!(simplified.value().as_expression().is_ok());
+/// - For non-partials, deep deref. TODO(ap/gj): deep deref.
+pub fn simplify_bindings(bindings: Bindings) -> Option<Bindings> {
+    let mut unsatisfiable = false;
+    let mut simplify = |var: Symbol, term: Term| {
+        let simplified = simplify_partial(&var, term);
+        match simplified.value().as_expression() {
+            Ok(o) if o == &FALSE => unsatisfiable = true,
+            _ => (),
+        }
+        simplified
+    };
 
-                (var, simplified)
+    let bindings: Bindings = bindings
+        .iter()
+        .map(|(var, value)| match value.value() {
+            Value::Expression(o) => {
+                assert_eq!(o.operator, Operator::And);
+                (var.clone(), simplify(var.clone(), value.clone()))
             }
-            _ => (var, value),
+            Value::Variable(v) | Value::RestVariable(v)
+                if v.is_temporary_var()
+                    && bindings.contains_key(v)
+                    && matches!(bindings[v].value(), Value::Variable(_) | Value::RestVariable(_)) =>
+            {
+                (var.clone(), bindings[v].clone())
+            }
+            _ => (var.clone(), value.clone()),
         })
-        .collect()
+        .collect();
+
+    if unsatisfiable {
+        None
+    } else {
+        Some(bindings)
+    }
+}
+
+pub struct Simplifier {
+    bindings: Bindings,
+    this_var: Symbol,
 }
 
-pub struct Simplifier;
 impl Folder for Simplifier {
-    /// Deduplicate constraints.
-    fn fold_partial(&mut self, partial: Partial) -> Partial {
-        fold_partial(self.deduplicate_constraints(partial), self)
+    fn fold_term(&mut self, t: Term) -> Term {
+        fold_term(self.deref(&t), self)
```

Comment:
This recurses infinitely for me with a sample policy. Reproed test case:
```
    #[test]
    fn test_partial_matches_with_cycle() -> TestResult {
        let p = Polar::new();
        p.load_str(
            r#"
                f(x) if g(x) and h(x);
                g(x) if x.foo = x;
                h(x: Integer);
            "#,
        )?;

        let mut q = p.new_query_from_term(term!(call!("f", [sym!("x")])), false);
        assert_query_done!(q); // ignoring for the sake of brevity what the result _should_ be
        Ok(())
    }
```

The use of matches is so that it hits the code path in `isa_expr` on vm.rs line 1252.